freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtRoleComponent.java

```diff
@@ -53,7 +53,18 @@ public void addRoleAndPrivileges(Optional<Service> service, Optional<Host> host,
         if (roleRequest != null && StringUtils.isNotBlank(roleRequest.getRoleName())) {
             Set<Role> allRole = ipaClient.findAllRole();
             Optional<Role> optionalRole = allRole.stream().filter(role -> role.getCn().equals(roleRequest.getRoleName())).findFirst();
-            Role role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            Role role;
+            try {
+                role = optionalRole.isPresent() ? optionalRole.get() : ipaClient.addRole(roleRequest.getRoleName());
+            } catch (FreeIpaClientException e) {
+                if (!FreeIpaClientExceptionUtil.isDuplicateEntryException(e)) {
+                    LOGGER.error("Failed to add the role [{}]", roleRequest.getRoleName(), e);
+                    throw e;
+                }
+                LOGGER.debug("The role [{}] was recently created in a different thread, retrieve it", roleRequest.getRoleName(), e);
+                allRole = ipaClient.findAllRole();
+                role = allRole.stream().filter(r -> r.getCn().equals(roleRequest.getRoleName())).findFirst().get();
```

Comment:
this looks like a duplicate of line 54-55. Could you move this to a common method?